Stopping informer does not cancel the resyncFuture nor it shutdown the resyncExecutor

While trying to stop a set of  `informers` and start new ones, there is remaining task trying to schedule in an executor:

```
2021-04-19 16:14:50 SEVERE  hudson.init.impl.InstallUncaughtExceptionHandler$DefaultUncaughtExceptionHandler uncaughtException A thread (informer-controller-ImageStream/460) died unexpectedly due to an uncaught exception, this may leave your Jenkins in a bad way and is usually indicative of a bug in the code.
```

Incidently with the error, I have several errors of this type showing that a task is still trying to be submitted on an open executor.

```
2021-04-19 16:15:01 WARNING io.fabric8.kubernetes.client.dsl.internal.WatcherWebSocketListener onFailure Exec Failure
java.io.InterruptedIOException: executor rejected
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:186)
	at okhttp3.Dispatcher.promoteAndExecute(Dispatcher.java:186)
	at okhttp3.Dispatcher.enqueue(Dispatcher.java:137)
	at okhttp3.RealCall.enqueue(RealCall.java:127)
	at okhttp3.internal.ws.RealWebSocket.connect(RealWebSocket.java:193)
[...]
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.RejectedExecutionException: Task okhttp3.RealCall$AsyncCall@6285bec1 rejected from java.util.concurrent.ThreadPoolExecutor@509c2579[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 15]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:183)
	... 104 more
```

After investigations, I see that the `resyncFuture`  https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java#L103  of informers are not canceled not their executors are shutdown.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ### 5.4-SNAPSHOT
 
 #### Bugs
+* Fix #3014: Resync Future is canceled and resync executor is shutdown on informer stop
 * Fix #2989: serialization will generate valid yaml when using subtypes
 * Fix #2996: Generating CRDs from the API should now properly work
 * Fix #3000: Set `no_proxy` in the okhttp builder in case the `proxy_url` is null

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
@@ -123,6 +123,8 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
     synchronized (this) {
       reflector.stop();
       reflectExecutor.shutdown();
+      resyncFuture.cancel(true);
+      resyncExecutor.shutdown();
     }
   }
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
@@ -91,16 +91,17 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
 
     // Starts one daemon thread for resync
     this.resyncExecutor = Executors.newSingleThreadScheduledExecutor();
-    initReflector();
+    this.reflector = new Reflector<>(apiTypeClass, listerWatcher, queue, operationContext, fullResyncPeriod);
   }
 
   public void run() {
     log.info("informer#Controller: ready to run resync and reflector runnable");
-
     // Start the resync runnable
     if (fullResyncPeriod > 0) {
       ResyncRunnable resyncRunnable = new ResyncRunnable(queue, resyncFunc);
-      resyncFuture = resyncExecutor.scheduleAtFixedRate(resyncRunnable, fullResyncPeriod, fullResyncPeriod, TimeUnit.MILLISECONDS);
+      if(!resyncExecutor.isShutdown()) {
+        resyncFuture = resyncExecutor.scheduleAtFixedRate(resyncRunnable, fullResyncPeriod, fullResyncPeriod, TimeUnit.MILLISECONDS);
+      }
     } else {
       log.info("informer#Controller: resync skipped due to 0 full resync period");
     }
@@ -143,9 +144,6 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
    * @return latest resource version
    */
   public String lastSyncResourceVersion() {
-    if (reflector == null) {
-      return "";
-    }
     return reflector.getLastSyncResourceVersion();
   }
 
@@ -169,10 +167,6 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
         throw e;
       }
     }
-  }
-
-  private void initReflector() {
-      reflector = new Reflector<>(apiTypeClass, listerWatcher, queue, operationContext, fullResyncPeriod);
   }
 
   ScheduledExecutorService getReflectExecutor() {

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Controller.java
@@ -123,7 +123,9 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
     synchronized (this) {
       reflector.stop();
       reflectExecutor.shutdown();
-      resyncFuture.cancel(true);
+      if (resyncFuture != null) {
+        resyncFuture.cancel(true);
+      }
       resyncExecutor.shutdown();
     }
   }
@@ -171,5 +173,13 @@ public class Controller<T extends HasMetadata, L extends KubernetesResourceList<
 
   private void initReflector() {
       reflector = new Reflector<>(apiTypeClass, listerWatcher, queue, operationContext, fullResyncPeriod);
+  }
+
+  ScheduledExecutorService getReflectExecutor() {
+    return this.reflectExecutor;
+  }
+
+  ScheduledExecutorService getResyncExecutor() {
+    return this.resyncExecutor;
   }
 }

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ControllerTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ControllerTest.java
@@ -15,24 +15,33 @@
  */
 package io.fabric8.kubernetes.client.informers.cache;
 
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+import org.mockito.Mockito;
+
 import io.fabric8.kubernetes.api.model.Pod;
 import io.fabric8.kubernetes.api.model.PodList;
 import io.fabric8.kubernetes.client.dsl.base.OperationContext;
 import io.fabric8.kubernetes.client.informers.ListerWatcher;
 import io.fabric8.kubernetes.client.informers.SharedInformerEventListener;
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
-import org.mockito.Mockito;
-
-import java.util.concurrent.ConcurrentLinkedQueue;
-
-import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertThrows;
 
 class ControllerTest {
   private DeltaFIFO<Pod> deltaFIFO = Mockito.mock(DeltaFIFO.class, Mockito.RETURNS_DEEP_STUBS);
-  private ListerWatcher<Pod, PodList> listerWatcher = Mockito.mock(ListerWatcher.class, Mockito.RETURNS_DEEP_STUBS);
+  private abstract class AbstractPodListerWatcher implements ListerWatcher<Pod, PodList> {};
+
+  private ListerWatcher<Pod, PodList> listerWatcher = Mockito.mock(AbstractPodListerWatcher.class, Mockito.RETURNS_DEEP_STUBS);
   private OperationContext operationContext = Mockito.mock(OperationContext.class, Mockito.RETURNS_DEEP_STUBS);
   private ConcurrentLinkedQueue<SharedInformerEventListener> eventListeners = Mockito.mock(ConcurrentLinkedQueue.class, Mockito.RETURNS_DEEP_STUBS);
 
@@ -82,9 +91,222 @@ class ControllerTest {
 
     // When
     controller.stop();
-
     // Then
     assertThat(controller.getReflectExecutor().isShutdown()).isTrue();
     assertThat(controller.getResyncExecutor().isShutdown()).isTrue();
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period should have synced")
+  void testControllerHasSync() throws InterruptedException {
+    // Given + When
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { }, () -> true,
+      10L, operationContext, eventListeners);
+    Thread controllerThread = newControllerThread(controller);
+    controllerThread.start();
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(controller.hasSynced()).isFalse();
+    assertThat(controller.lastSyncResourceVersion()).isNull();
+  }
+
+  @Test
+  @DisplayName("Controller with interrupted thread should not shutdown resyncExecutor")
+  void testControllerRunWithInterruptedThread() throws InterruptedException {
+    // Given + When
+    // used to be able to interrupt the thread in the lambda
+    ThreadWrapper controllerThreadWrapper = new ThreadWrapper(); 
+    long fullResyncPeriod = 1L;
+    int numberOfResyncs = 1;
+    final CountDownLatch countDown = new CountDownLatch(numberOfResyncs);
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> {},
+      () -> {
+          controllerThreadWrapper.interrupt();
+        return true;
+      },
+      fullResyncPeriod, operationContext, eventListeners);
+    Thread controllerThread = newControllerThread(controller);
+    controllerThreadWrapper.thread = controllerThread; // put the thread in the wrapper, so the lamba can interrupt it
+    controllerThread.start();
+    countDown.await(500, TimeUnit.MILLISECONDS); // a too short value does not allow the processLoop to start.
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isFalse();
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period should initialize resyncExecutor")
+  void testControllerRunWithResyncPeriodGreaterThanZero() throws InterruptedException {
+    // Given + When
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> true,
+      1L, operationContext, eventListeners);
+    Thread controllerThread = newControllerThread(controller);
+    controllerThread.start();
+    controller.stop();
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isTrue();
+  }
+
+  @Test
+  @DisplayName("Controller with resync function throwing exception")
+  void testControllerRunsReyncFunctionThrowingException() throws InterruptedException {
+    // Given + When
+    long fullResyncPeriod = 10L;
+    int numberOfResyncs = 10;
+    final CountDownLatch countDown = new CountDownLatch(numberOfResyncs);
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> {
+        countDown.countDown();
+        if( countDown.getCount() == 2 ) {
+            RuntimeException exception = new RuntimeException("make it fail");
+          throw exception;
+        }
+        return true;
+      },
+      fullResyncPeriod, operationContext, eventListeners);
+
+    Executable controllerRun = newControllerRun(controller);
+    assertDoesNotThrow(controllerRun);
+    // We give an extra cycle to avoid clock inaccurracy interruptions
+    countDown.await(( numberOfResyncs + 1 )  * fullResyncPeriod, TimeUnit.MILLISECONDS);
+    controller.stop();
+    // Then
+    assertThat(countDown.getCount()).isLessThanOrEqualTo(2);
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period should initialize resyncExecutor")
+  void testControllerRunWithResyncPeriodGreaterThanZeroAndExecutorNotShutdown() throws InterruptedException {
+    // Given + When
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> true,
+      1L, operationContext, eventListeners);
+    Executable controllerRun = newControllerRun(controller);
+    assertDoesNotThrow(controllerRun);
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    // Then
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isFalse();
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period should initialize resyncExecutor")
+  void testControllerRunWithResyncPeriodGreaterThanZeroAndExecutorForcedShutdown() throws InterruptedException {
+    // Given + When
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> true,
+      1L, operationContext, eventListeners);
+    Executable controllerRun = newControllerRun(controller);
+    assertDoesNotThrow(controllerRun);
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    // Then
+    assertNotNull(resyncExecutor);
+    resyncExecutor.shutdown();
+    assertThat(resyncExecutor.isShutdown()).isTrue();
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period to 0 should initialize resyncExecutor")
+  void testControllerRunWithResyncPeriodToZero() throws InterruptedException {
+    // Given + When
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> true,
+      0L, operationContext, eventListeners);
+    Thread controllerThread = newControllerThread(controller);
+    controllerThread.start();
+    controller.stop();
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isTrue();
+    assertThat(resyncExecutor.isTerminated()).isTrue();
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period should run, initialize resyncExecutor and resync at least a given number of times")
+  void testControllerRunsReyncFunctionExpectedNumberOfTime() throws InterruptedException {
+    // Given + When
+    long fullResyncPeriod = 10L;
+    int numberOfResyncs = 100;
+    final CountDownLatch countDown = new CountDownLatch(numberOfResyncs);
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> {countDown.countDown(); return true;},
+      fullResyncPeriod, operationContext, eventListeners);
+
+    Executable controllerRun = newControllerRun(controller);
+    assertDoesNotThrow(controllerRun);
+    // We give an extra cycle to avoid clock inaccurracy interruptions
+    countDown.await(( numberOfResyncs + 1 )  * fullResyncPeriod, TimeUnit.MILLISECONDS);
+    controller.stop();
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isTrue();
+    assertThat(resyncExecutor.isTerminated()).isTrue();
+    assertThat(countDown.getCount()).isLessThanOrEqualTo(1);
+  }
+
+  @Test
+  @DisplayName("Controller initialized with resync period to 0 should run but never resync")
+  void testControllerNeverRunsReyncFunctionWhenPeriodIsZero() throws InterruptedException {
+    // Given + When
+    int count = 10;
+    final CountDownLatch countDown = new CountDownLatch(count);
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> {countDown.countDown(); return true;},
+      0, operationContext, eventListeners);
+    Executable controllerRun = newControllerRun(controller);
+    assertDoesNotThrow(controllerRun);
+    // We give an extra cycle to avoid clock inaccurracy interruptions
+    countDown.await(1000, TimeUnit.MILLISECONDS);
+    controller.stop();
+    // Then
+    ScheduledExecutorService resyncExecutor = controller.getResyncExecutor();
+    assertNotNull(resyncExecutor);
+    assertThat(resyncExecutor.isShutdown()).isTrue();
+    assertThat(countDown.getCount()).isEqualTo(count);
+  }
+
+  private class ThreadWrapper{
+    public Thread thread;
+    public void interrupt() {
+      if( thread != null) {
+        thread.interrupt();
+      }
+    }
+  }
+
+  private Executable newControllerRun(Controller<Pod, PodList> controller) {
+    return new Executable() {
+      @Override
+      public void execute() {
+        Thread controllerThread = newControllerThread(controller);
+        controllerThread.start();
+      }
+    };
+  }
+
+  private Thread newControllerThread(Controller<Pod, PodList> controller) {
+    Thread controllerThread = new Thread( new Runnable() {
+      @Override
+      public void run() {
+        controller.run();
+      }
+    });
+    return controllerThread;
   }
 }

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ControllerTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ControllerTest.java
@@ -26,6 +26,7 @@ import org.mockito.Mockito;
 
 import java.util.concurrent.ConcurrentLinkedQueue;
 
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
@@ -68,5 +69,22 @@ class ControllerTest {
 
     // Then
     assertEquals(0L, controller.getReflector().getResyncPeriodMillis());
+  }
+
+  @Test
+  @DisplayName("Controller stop shut downs/cancels all executor services")
+  void testStop() {
+    // Given
+    Controller<Pod, PodList> controller = new Controller<>(Pod.class, deltaFIFO, listerWatcher,
+      simpleEntries -> { },
+      () -> true,
+      1000L, operationContext, eventListeners);
+
+    // When
+    controller.stop();
+
+    // Then
+    assertThat(controller.getReflectExecutor().isShutdown()).isTrue();
+    assertThat(controller.getResyncExecutor().isShutdown()).isTrue();
   }
 }

